open-craft/edx-platform
# Adaptive Content Block: Send results to external service when learner submits review question

This PR adds logic for notifying Domoscio when a learner submits a review question.

cf. [OC-2152](https://tasks.opencraft.com/browse/OC-2152)

**Implementation**

When a learner submits a review question, ACB needs to send an `EventResult` event for the "knowledge node student" object that links the learner to the review question to Domoscio. The event must include information about the result that the learner obtained (pass/fail).

To make this possible, this PR introduces a new `TRACKING_BACKEND` called `AdaptiveLearningBackend`. Like other existing backends, the `AdaptiveLearningBackend` listens to events that the platform generates when users interact with courses. When it receives a [`problem_check` event](https://edx.readthedocs.io/projects/devdata/en/latest/internal_data_formats/tracking_logs.html#problem-check), it extracts relevant information from it and passes it to `AdaptiveLibraryContentModule`, which takes care of sending an appropriate event to Domoscio.

**Test instructions**

1. Follow instructions on the [epic](https://tasks.opencraft.com/browse/OC-2127) to set up a FUN box.

2. Switch to this branch (`adaptive-content-block`) and install the new block via

    ```sh
    pip install -r requirements/edx/local.txt
    ```

3. Create a new content library and add a few problems to it.

4. Create a new course and add `"adaptive_library_content"` to advanced modules.

5. Add an "Adaptive Content Block" that references the library created previously to a unit (optionally modifying the number of components to display to each student), and publish the unit. Do **not** access the unit in the LMS yet.

6. To be able to work with the Domoscio API, you'll need to specify a few course-level settings: In Studio, navigate to the Advanced Settings page for the course ("Paramètres" > "Paramètres avancés") and replace the default value for the "Adaptive Learning Configuration" setting with the following:

    ```json
    "url": "<see comment on internal ticket>",
    "instance_id": <see comment on internal ticket>,
    "access_token": "<see comment on internal ticket>",
    "api_version": "v1"
    ```

7. In a venv that has the `requests` library installed, create a knowledge node that corresponds to the unit you created earlier (`block_id` needs to be a 32-character hash), and review the results:

    ```python
    import requests, json
    response = requests.post("<url>/v1/instances/<instance_id>/knowledge_nodes", headers={"Authorization": "Token token=<access_token>"}, data={"knowledge_graph_id": <see comment on internal ticket>, "name": "Unit node", "uid": "<block_id of unit>"})
    json.loads(response.content)
    ```

    Then create a knowledge node for each of the problems that belong to the library you created earlier. Note that problems belonging to a content library are assigned unique block IDs for each Adaptive Content Block that references them; you'll need to use these context-specific IDs when creating corresponding knowledge nodes. Block IDs for problems are 20-character hashes.

8. Review the list of events that exist on the Domoscio test instance:

    ```python
    events = requests.get("<url>/v1/instances/<instance_id>/events", headers={"Authorization": "Token token=<access_token>"})
    json.loads(events.content)
    ```

9. Access the unit that contains the ACB in the LMS. It will display all of the child blocks belonging to the ACB. *This is just for testing purposes.* (You won't want to wait until reviews become available :)) I'll remove the [line](https://github.com/open-craft/edx-platform/pull/50/files#diff-5b223f1aa12d0cba135cf45dade27de3R555) that disables the selection logic introduced in #49 before merging this PR.

10. Review the list of events as described above. It should now contain one more event (this is a "Unit viewed" event that gets sent when accessing a unit containing an ACB in the LMS; it should reference the "knowledge node student" object linking the unit you created earlier to the student that you used to access the unit in the LMS).

11. Answer one of the review questions and submit it (click "Vérifier").

12. Review the list of events as described above. It should now contain one more event. This is the result event that notified Domoscio about the fact that you submitted a review question, and what the results were. It should reference the "knowledge node student" object linking the review question you submitted to the current user, and its `"payload"` property should be set to the value corresponding to the result you obtained for the review question (`"0"` if you got the question wrong, `"100"` if you got it right).

**Relevant tests**

Before running any of the commands listed below, remove (or comment out) the [line](https://github.com/open-craft/edx-platform/pull/50/files#diff-5b223f1aa12d0cba135cf45dade27de3R555) that disables the selection logic introduced in #49.

The commands for running relevant tests are:

```sh
paver test_system -t lms/djangoapps/course_blocks/transformers/tests/test_library_content.py --fasttest  # Didn't make changes to this file, but the tests it defines should still be passing
paver test_lib -l common/lib/xmodule/xmodule/tests/test_library_content.py
paver test_lib -l common/lib/xmodule/xmodule/tests/test_util_adaptive_learning.py
paver test_system -t common/djangoapps/track/backends/tests/test_adaptive_learning.py --fasttest
```

**Reviewers**

- [ ] @e-kolpakov / @pomegranited

**Notes**

- The choice of result values that get sent to Domoscio on pass/fail (`"100"` and `"0"`, respectively) is based on information from the Domoscio API documentation. I'm just mentioning this here because I was surprised initially that the values had to be sent as strings.
